### PR TITLE
[FEATURE] Add waiting list

### PR DESCRIPTION
**Pull request by jnguiot**
Ajout d'un système de liste d'attente.
https://trello.com/c/xzqs6BiX/49-g%C3%A9rer-une-liste-dattente
Je n'ai pas fait le système de vérification d'inscription unique, car je pense que ça doit être discuté plus largement (cf notre discussion de lundi)

### CODE REVIEW

**Review by gdaviet — CHANGES_REQUESTED** [review]
Joli travail !

 - Le request change c'est surtout pour le coup du has_free_slots() vs has_free_online_slots(), AMA on  ne devrait utiliser que le second
 - Question: Que se  passe-t-il si l'encadrant augmente le nombre de places de la collective ? Faudrait-il d'abord vider la liste d'attente ?

File: collectives/models/registration.py [hunk 1]
```diff
@@ -73,13 +77,19 @@ def transition_table(cls, requires_payment):
...
+                cls.Waiting,
+            ],
+            cls.Rejected: [re_register_status, cls.ToBeDeleted, cls.Waiting],
             cls.PaymentPending: [cls.Rejected],
-            cls.SelfUnregistered: [re_register_status, cls.ToBeDeleted],
+            cls.SelfUnregistered: [re_register_status, cls.Waiting, cls.ToBeDeleted],
             cls.JustifiedAbsentee: [cls.Rejected, cls.Active, cls.UnJustifiedAbsentee],
             cls.UnJustifiedAbsentee: [cls.Rejected, cls.Active, cls.JustifiedAbsentee],
             cls.ToBeDeleted: [],
+            cls.Waiting: [cls.Rejected, re_register_status],
```
[the remark this answers is not in the document]
Comment (jnguiot) [reply collectives/models/registration.py:92]: Le passage de Waiting à Active est autorisé par `re_register_status`, non?

File: collectives/templates/partials/event/self_register.html [hunk 2]
```diff
@@ -1,5 +1,8 @@
+{% macro button(waiting_list = False) -%}
+
+
 <p>Les inscriptions en ligne sont possibles jusqu'au {{format_datetime(event.registration_close_time)}}. </p>
-{% if event.requires_payment() %}
+{% if event.requires_payment() and not waiting_list %}
```
[the remark this answers is not in the document]
Comment (jnguiot) [reply collectives/templates/partials/event/self_register.html:5]: Je vais faire le https://trello.com/c/5qdqybDf/256-afficher-les-tarifs-dune-collective-pour-lencadrant-et-ladh%C3%A9rent-lorsque-les-inscriptions-ne-sont-pas-ouvertes dans la foulée

**Comment by jnguiot** [comment]
> Question: Que se passe-t-il si l'encadrant augmente le nombre de places de la collective ? Faudrait-il d'abord vider la liste d'attente ?

Pour moi, oui: en cas d'augmentation du nombre de place, la file d'attente doit être "consommée". En gros, il faut je fasse un `update_waiting_list()` après la modification d'un event.

File: collectives/models/registration.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gdaviet) [reply collectives/models/registration.py:92]: Ah oui effectivement, j'avais zappé ça !

File: collectives/templates/partials/event/self_register.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (gdaviet) [reply collectives/templates/partials/event/self_register.html:5]: Justement cette partie de la carte est déjà implémentée, je pense qu'on peut utiliser le même code.
La partie qui manque c'est la prévisualisation de tous les tarifs pour les encadrants/modérateurs, mais ça je pense que ca se passe plus sur la page d'édition des tarifs (et c'est un peu complexe niveau UI car il faut prévoir toutes les combinaisons de type de licence, encadrant ou non, etc)

[the remark this answers is not in the document]
Comment (jnguiot) [reply collectives/templates/partials/event/self_register.html:5]: Elle est déjà implémentée?
To be honest, je suis en train de travailler dessus pour afficher tous les tarifs tout le temps en bas de la page (même s'ils ne sont pas lié à la catégorie de l'utilisateur)

[the remark this answers is not in the document]
Comment (gdaviet) [reply collectives/templates/partials/event/self_register.html:5]: Mais effectivement on peut aussi inclure un tableau récapitulatif de tous les tarifs "pour information", même si ce sera différent de ce que les utilisateurs verront au moment de payer

**Comment by jnguiot** [comment]
@gdaviet Est-ce que tu peux me relire la table des prix stp?

File: collectives/api/payment.py [hunk 3]
```diff
@@ -174,6 +203,33 @@ def list_payments(event_id=None):
...
+    """
+    event = Event.query.get(event_id)
+    if event is None:
+        return abort(404)
+    if not event.requires_payment():
+        return abort(400)
+
+    result = (
+        ItemPrice.query.filter_by(enabled=True)
+        .filter(PaymentItem.event_id == event_id)
```
Comment (gdaviet) [inline collectives/api/payment.py:223]: Il manque un `ItemPrice.item_id == PaymentItem.id` ou quelque chose du genre non? J'ai l'impression que ca m'affiche tous les tarifs, pas juste ceux de la collective

File: collectives/templates/partials/event/payment_item_list.html [hunk 4]
```diff
@@ -0,0 +1,10 @@
+
+  <script type="text/javascript" src="https://unpkg.com/tabulator-tables@4.9.3/dist/js/tabulator.js"></script>
+  <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.24.0/moment-with-locales.min.js"></script>
+  <script type="text/javascript" > const event_id = {{event.id}} ;</script>
+  <script type="text/javascript" src="{{ url_for('static', filename='js/event/payment_item_list.js') }}"></script>
```
Comment (gdaviet) [inline collectives/templates/partials/event/payment_item_list.html:5]: Préférence personnelle pour passer l'event_id comme argument d'une fonctioon createTable() plutot que d'introduire une variable globale

**Review by gdaviet — COMMENTED** [review]
Je suis un peu mitigé sur la table, je ne sais pas trop le nombre de tarifs crées par les encadrants en vrai. Là sur mes tests la table ca fait vraiment massif et il est assez difficile de se rendre compte quels tarifs seront vraiment proposés, mais il faut dire que c'est une configuration vraiment compliquée (et que la table contient des tarifs a priori d'autres événements), donc je pense que ce sera moins le cas en vrai.
Je serais quand même personnellement pour passer la table en dessous de la zone Administration, et ajouter la preview des tarifs personallisée pour la waiting list.

File: collectives/templates/editevent.html [hunk 5]
```diff
@@ -166,6 +166,9 @@ <h4 class="heading-4">Informations</h4>
     <div class="controls">
       <label for="num_slots">Nombre de participants : </label> {{ form.num_slots(size=4) }}
     </div>
+    <div class="controls">
```
Comment (fberu) [inline collectives/templates/editevent.html:169]: Réflexion: ne devrions-nous pas commencer à avoir un panneau avec les données optionnels ? Pas sûr que grand monde utilise cette option

File: collectives/templates/event.html [hunk 6]
```diff
@@ -226,6 +233,26 @@ <h3 class="heading-3"> Inscrits {{event.num_taken_slots()}} / {{event.num_slots}
         {% elif current_time > event.registration_close_time %}
         <p>Les inscriptions sont closes.</p>
         {% endif %}
+
```
Comment (fberu) [inline collectives/templates/event.html:236]: Lorsque la liste des inscrits est complètes et qu'une liste d'attente est possible, il faudrait afficher des messages expliquant cette liste d'attente:

`Le nombre maximum d'inscrits par internet a été atteint.` -> `Le nombre maximum d'inscrits par internet a été atteint. Vous pouvez cependant vous inscrire sur liste d'attente`

Dans ce cas là, le bouton `s'inscrire` devrait plus être `s'incrire sur liste d'attente` afin qu'il n'y ait aucune confusion

File: collectives/email_templates.py [hunk 7]
```diff
@@ -183,3 +183,34 @@ def send_cancelled_event_notification(name, event):
             )
     except BaseException as err:
         current_app.logger.error(f"Mailer error: {err}")
+
+
+def send_update_waiting_list_notification(registration):
```
Comment (fberu) [inline collectives/email_templates.py:188]: Tip top!

**Comment by fberu** [comment]
Super taf!
J'ai mis des commentaires sur des détails.

Après des tests et utilisation, se posent plusieurs questions:

* effectivement il y a le point d'inscription unique. Pour moi aucun soucis pour qu'une personne s'inscrire sur plusieurs listes d'attente le même jour MAIS que se passe-t-il s'il retrouve inscrit (pour de vrais) sur 2 collectives le même jour ? Il faudrait donc résoudre #427 et mettre en place la mécanique: si inscrit sur plusieurs d'attente, dès qu'il sort d'une, alors il sort de toute
* être averti par mail est top. Malheureusement les encadrants ont constaté que des gens s'auto désinscrivent le soir pour le lendemain (des fois à 2h du mat pour un départ à 7h). Dans ce cas là, celui sur liste d'attente est prévenu tardivement qu'il est inscrit et a de grandes chances de ne pas voir la notification. Et du coup l'encadrant grognera le matin au parking parce qu'il n'est pas là

Je pense qu'il y a aussi d'autres corner cases, compliqué à voir sans utilisation.

Ma proposition:

* fixer #427 
* lancer cette fonctionnalité en mode "beta" ce printemps / été afin d'avoir un max de feedbacks pour cet hiver: en effet c'est surtout le ski de rando qui a besoin de cette fonctionnalité

**Comment by fberu** [comment]
J'ai oublié: il serait bien d'avoir sur son profil les collectives sur lesquelles un adhérent est en liste d'attente

File: collectives/templates/event.html (hunk 6 above)
Comment (jnguiot) [reply collectives/templates/event.html:236]: > Le nombre maximum d'inscrits par internet a été atteint. -> Le nombre maximum d'inscrits par internet a été atteint. Vous pouvez cependant vous inscrire sur liste d'attente

Ca me parait assez explicite pourtant, vu que la liste d'attente est affichée juste en dessous :thinking:

> Dans ce cas là, le bouton s'inscrire devrait plus être s'incrire sur liste d'attente afin qu'il n'y ait aucune confusion

Je vais rajouter ça, pour être sur que les adhérents voient que c'est une liste d'attente

File: collectives/templates/editevent.html (hunk 5 above)
Comment (jnguiot) [reply collectives/templates/editevent.html:169]: Pourquoi pas, à voir avec @gregone . Cependant, j'aime bien afficher toutes les options quand même: ca permet aux encadrants de voir toutes les options proposées.
Je pense rester comme ça pour cette PR, mais on pourra modifier par la suite.

File: collectives/api/payment.py (hunk 3 above)
Comment (jnguiot) [reply collectives/api/payment.py:223]: Corrigé

File: collectives/templates/partials/event/payment_item_list.html (hunk 4 above)
Comment (jnguiot) [reply collectives/templates/partials/event/payment_item_list.html:5]: J'ai fait sauter cette partie

**Comment by jnguiot** [comment]
> J'ai oublié: il serait bien d'avoir sur son profil les collectives sur lesquelles un adhérent est en liste d'attente

vu

> Je suis un peu mitigé sur la table, je ne sais pas trop le nombre de tarifs crées par les encadrants en vrai. Là sur mes tests la table ca fait vraiment massif et il est assez difficile de se rendre compte quels tarifs seront vraiment proposés, mais il faut dire que c'est une configuration vraiment compliquée (et que la table contient des tarifs a priori d'autres événements), donc je pense que ce sera moins le cas en vrai. Et peut être que juste diminuer la taille de la police suffirait. On peut toujours merger et voir à l'usage sur test.collectives.

J'ai fait une table moins massive

> Je pense qu'il y a aussi d'autres corner cases, compliqué à voir sans utilisation.
> lancer cette fonctionnalité en mode "beta" ce printemps / été afin d'avoir un max de feedbacks pour cet hiver: en effet c'est surtout le ski de rando qui a besoin de cette fonctionnalité

En effet, il faut lancer et voir ce que ca donne.
